core/trino-main/src/main/java/io/trino/server/configuration/ConfigurationResolverManager.java

```diff
@@ -0,0 +1,172 @@
+    private final AtomicReference<Map<String, ConfigurationValueResolver>> configurationProviders = new AtomicReference<>(ImmutableMap.of());
+    private final File installedConfigurationResolver;
+    private final List<File> configFiles;
+
+    @Inject
+    public ConfigurationResolverManager(ConfigurationResolverConfig config)
+    {
+        this.installedConfigurationResolver = config.getInstalledConfigurationResolver();
+        configFiles = ImmutableList.copyOf(config.getConfigurationResolverFiles());
+    }
+
+    public void load()
+    {
+        listFiles(installedConfigurationResolver).stream()
+                .filter(File::isDirectory)
+                .forEach(file -> loadConfigurationResolvers(file.getAbsolutePath(), () -> createClassLoader(file.getName(), buildClassPath(file))));
+        ImmutableMap.Builder<String, ConfigurationValueResolver> builder = ImmutableMap.builder();
+        for (File configFile : configFiles) {
+            builder.put(getNameWithoutExtension(configFile.getName()), loadConfigProvider(configFile.getAbsoluteFile()));
+        }
+        this.configurationProviders.set(builder.buildOrThrow());
+    }
+
+    private static List<File> listFiles(File path)
+    {
+        try {
+            try (DirectoryStream<Path> directoryStream = newDirectoryStream(path.toPath())) {
+                return stream(directoryStream)
+                        .map(Path::toFile)
+                        .sorted()
+                        .collect(toImmutableList());
+            }
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private void loadConfigurationResolvers(String plugin, Supplier<PluginClassLoader> createClassLoader)
+    {
+        log.info("-- Loading plugin %s --", plugin);
+
+        PluginClassLoader pluginClassLoader = createClassLoader.get();
+
+        log.debug("Classpath for plugin:");
+        for (URL url : pluginClassLoader.getURLs()) {
+            log.debug("    %s", url.getPath());
+        }
+
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(pluginClassLoader)) {
+            loadPlugin(pluginClassLoader);
+        }
+
+        log.info("-- Finished loading plugin %s --", plugin);
+    }
+
+    public ConfigurationResolver getConfigurationResolver()
+    {
+        return new ConfigurationResolver(configurationProviders.get());
+    }
+
+    public void addConfigurationProviderFactory(ConfigurationValueResolverFactory configurationValueResolverFactory)
+    {
+        configurationProviderFactories.put(configurationValueResolverFactory.getName().toLowerCase(ENGLISH), configurationValueResolverFactory);
+    }
+
+    private void loadPlugin(PluginClassLoader pluginClassLoader)
+    {
+        ServiceLoader<ConfigurationValueResolverFactory> serviceLoader = ServiceLoader.load(ConfigurationValueResolverFactory.class, pluginClassLoader);
```

Comment:
Should we introduce a new Plugin like interface - which would return a `List<ConfigurationValueResolverFactory>` so that a same plugin would allow us to return multiple factories - like we do for `trino-password-authenticators`